app/models/moved_post.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class MovedPost < ActiveRecord::Base
+  belongs_to :old_topic, class_name: "Topic", foreign_key: :old_topic_id
+  belongs_to :old_post, class_name: "Post", foreign_key: :old_post_id
+
+  belongs_to :new_topic, class_name: "Topic", foreign_key: :new_topic_id
+  belongs_to :new_post, class_name: "Post", foreign_key: :new_post_id
+end
+
+# == Schema Information
+#
+# Table name: moved_posts
+#
+#  id                :bigint           not null, primary key
+#  old_topic_id      :bigint
+#  old_post_id       :bigint
+#  old_post_number   :bigint
```

Comment:
Do you plan to add code in a follow-up PR that sets these columns to `NULL` when the old topic or post gets deleted?

Reply:
Oh interesting, I hadn't considered that. I was actually curious how these columns worked, as the old and new IDs seems to be the same in testing.. So I guess after the moving is done, we don't need to reference `old_` anything?

Reply:
Well thinking about it.. I think I'm OK with referencing post/topic IDs that are deleted here.. I'm not sure how imporant that part is. There is no usage that would cause errors/downtime as-is. I'm not sure if it's necessary or not.

Reply:
I was only wondering because I noticed that those columns are nullable. If you don't intend to set those columns `NULL` then I think it's best to create them as `NOT NULL`, or am I missing something.

> I was actually curious how these columns worked, as the old and new IDs seems to be the same in testing...

It's a long time since I wrote that code, but if memory serves, it works like this:

* `old_topic_id` and `new_topic_id` should always be different
* `old_post_id` and `new_post_id` should be equal unless it's the first post in a topic
* `old_post_number` and `new_post_number` might be the same, but usually isn't, depends on the target topic

I think it's good practice to clean up tables if stuff gets hard deleted. Maybe add relations to the `Topic` and `Post` model and mark them as `dependent: :destroy`? Or handle it i `PostDestroyer` like we do for other post related tables.

https://github.com/discourse/discourse/blob/e82e255531b4e17ab06a92bfb3cd2cc61074658a/lib/post_destroyer.rb#L186-L190

Reply:
Thanks for the details! I do worry though, about losing data. Let's say we move posts from topicX -> topicY, and then delete topic. I still want to know that those posts on topicY came from _somewhere else_, even if the original topic is deleted.

I think that is why I made them nullable. Maybe as you say in `PostDestroyer` we `null` those values out (deleted post/topic_ids) but keep the records around. 

Does that make sense to you?

Reply:
I agree. It's good to not lose the "this post was moved" metadata. That's why I'm very torn on the solution here.

Option 1: Make the `old_*` columns `NOT NULL` and don't do anything when topics/posts are destroyed
Option 2: Set `old_columns` to `NULL` when the old topic or post is destroyed. Delete `moved_posts` when the new topic or post is destroyed.

Maybe it's best to go with the simplest solution, option 1. :thinking:  As you said, there's no usage that would cause errors. There's no guarantee that the `Topic` or `Post` models are actually used for deletion, so the data in `moved_posts` can't be fully trusted anyway. Yeah, I'm leaning toward option 1. Sorry for sending you in circles.